### DISCUSSION TITLE
Cannot save a file with FileSaver on Google Drive directly

### DISCUSSION THREAD

**Discussion by FrankPohl** [CommunityToolkit/Maui#1939] [Q&A]
I am using the FileSaver component to write small text files. This works fine one local folders but when I select a location in Drive I always get an error.
Error is:
 Unable to resolve absolute path where the file was saved 'content://com.google.android.apps.docs.storage/document/acc%3D4%3Bdoc%3Dencoded%

In fact I selected a folder and the file is created locally but empty.
Do I need to set a permission or what is the problem here?
Thanks frank

**Comment by bijington** [comment]
Did you add the following permissions as detailed in our docs: https://learn.microsoft.com/en-us/dotnet/communitytoolkit/maui/essentials/file-saver?tabs=android

If so and it's still not working have you looked through the linked Android documentation page: https://developer.android.com/reference/android/Manifest.permission#WRITE_EXTERNAL_STORAGE

**Reply by FrankPohl** [reply]
Yes, I have checked all the necessary things.
Meanwhile I used the FileSaver sample that is provided in this repository. 
It throws the same error when I select Google Drive as  a destination to save a file.
"Unable to resolve absolute path where the file was saved 'content://com.google.android.apps.docs.sto…"
I tried this on Android Version 14 and on Android Version 10 and it is the same on both versions

**Reply by bijington** [reply]
What if you tried just using something like `File.WriteAllText("abc", path);`

Where path is that location from your error? Do you get the same error?

**Reply by FrankPohl** [reply]
I tried this in the Visual Studio immediate window after the check for the save result.
`File.WriteAllText(@"content://com.google.android.apps.docs.storage/document/acc%3D4%3Bdoc%3Dencoded%3DKu881aYlpoT0MUenPmIEK5Y4Al-bvziy7JXrVoJyyIrTpC4HP7s%3D", "abc")`

This call raised gives this error message:
System.IO.DirectoryNotFoundException: Could not find a part of the path `'/content:/com.google.android.apps.docs.storage/document/acc%3D4%3Bdoc%3Dencoded%3DKu881aYlpoT0MUenPmIEK5Y4Al-bvziy7JXrVoJyyIrTpC4HP7s%3D'.`

**Reply by beefydog** [reply]
Does anyone know if it's even possible to save to a Google Drive via this toolkit or ANY other filesaving library or method? I've never gotten this to work either. Permissions set correctly, can save to any local device. I got this to work with the Android IDE and Kotlin, but for MAUI, no joy.

**Reply by VladislavAntonyuk** [reply]
You should work with Google drive api to save your files. Toolkit provides you a way to save files locally and is limited only by Google. 
If you can write with kotlin just migrate your code to c#.

**Reply by beefydog** [reply]
Kind of defeats the purpose of using MAUI (for mobile platforms). If Google is limiting this, then Kotlin or Java or any platform wouldn't work (and there's a ton of Android apps that can save to remote places like Drive) - so that cannot be true. The limitation is in this library, not Google

**Reply by bijington** [reply]
If it can be done in Kotlin or Java then it can be done in MAUI. This post https://stackoverflow.com/questions/12164024/android-open-and-save-files-to-from-google-drive-sdk#12321331 points out how to achieve it on Android. It all recommends that you use the Google Drive SDK as the others in this thread have suggested

**Comment by GeorgeLeithead** [comment]
I suggest you have a read of this article [Upload file to Google](https://www.daimto.com/upload-file-to-google-drive/).  It's not straight forward as you'll need all kinds of authentication and authorisation, etc.  You may also want to have a look at the [Google Docs API Client Library for .NET](https://developers.google.com/api-client-library/dotnet/apis/docs/v1).

**Reply by beefydog** [reply]
I ran into this issue, too. Permission settings in the Android Manifest are not enough. I'm surprised the community toolkit has not addressed this.

**Comment by daigs** [comment]
public static class UriExtensions
{
    /// <summary>
    /// Convert <see cref="Uri"/> to an absolute path (<see cref="string"/>).
    /// It returns <see langword="null"/> if the <see cref="Uri"/> is not resolvable.
    /// </summary>
    public static string? ToPhysicalPathX(this Uri uri)
    {
        if (uri == null)
        {
            return null;
        }

        var isAfter44 = Build.VERSION.SdkInt >= BuildVersionCodes.Kitkat;

        if (isAfter44 && IsDocumentUri(Application.Context, uri))
        {
            var path = GetPathFromDocumentUriAsync(Application.Context, uri);
            return path;
        }

        if (uri.Scheme == "content")
        {
            var path = QueryAbsolutePathAsync(Application.Context, uri);
            return path;
        }

        if (uri.Scheme == "file")
        {
            var path = uri.Path;
            return path;
        }

        return null;
    }

    private static string? GetPathFromDocumentUriAsync(Context context, Uri uri)
    {
        var authority = uri.Authority!;

        if (authority == "com.android.externalstorage.documents")
        {
            var docId = GetDocumentId(uri);
            var parts = docId.Split(':');
            if (parts.Length == 2)
            {
                var type = parts[0];
                if (type.Equals("primary", StringComparison.OrdinalIgnoreCase))
                {
                    var rootPath = Path.Combine(Environment.ExternalStorageDirectory!.Path, parts[1]);
                    return rootPath;
                }
                else
                {
                    var rootPath = Path.Combine("/storage", type, parts[1]);
                    return rootPath;
                }
            }
        }
        else if (authority == "com.android.providers.downloads.documents")
        {
            var docId = GetDocumentId(uri);
            if (docId.StartsWith("raw:"))
            {
                return docId.Substring(4);
            }

            var downloadId = long.Parse(docId);
            var downloadUri = ContentUris.WithAppendedId(Uri.Parse("content://downloads/public_downloads")!, downloadId);
            return QueryAbsolutePathAsync(context, downloadUri);
        }
        else if (authority == "com.android.providers.media.documents")
        {
            var docId = GetDocumentId(uri);
            string[] parts = docId.Split(':');
            if (parts.Length == 2)
            {
                var type = parts[0];
                var id = long.Parse(parts[1]);

                var mediaUri = (type switch
                {
                    "image" => MediaStore.Images.Media.ExternalContentUri,
                    "video" => MediaStore.Video.Media.ExternalContentUri,
                    "audio" => MediaStore.Audio.Media.ExternalContentUri,
                    _ => null
                })!;

                if (mediaUri != null)
                {
                    mediaUri = ContentUris.WithAppendedId(mediaUri, id);
                    return QueryAbsolutePathAsync(context, mediaUri);
                }
            }
        }

        return null;
    }

    private static string GetDocumentId(Uri uri)
    {
        return DocumentsContract.GetDocumentId(uri)!;
    }

    private static string? QueryAbsolutePathAsync(Context context, Uri uri)
    {
        string[] projection = [MediaStore.MediaColumns.Data];
        using var cursor = context.ContentResolver.Query(uri, projection, null, null, null);
        if (cursor != null && cursor.MoveToFirst())
        {
            var columnIndex = cursor.GetColumnIndex(MediaStore.MediaColumns.Data);
            if (columnIndex != -1)
            {
                return cursor.GetString(columnIndex);
            }
        }

        return null;
    }

    private static bool IsDocumentUri(Context context, Uri uri)
    {
        return DocumentsContract.IsDocumentUri(context, uri);
    }
}
